x-pack/legacy/plugins/maps/server/plugin.js

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+import { APP_ID, APP_ICON, createMapPath, MAP_SAVED_OBJECT_TYPE } from '../common/constants';
+import { watchStatusAndLicenseToInitialize } from '../../../server/lib/watch_status_and_license_to_initialize';
+import { initRoutes } from './routes';
+import { getEcommerceSavedObjects } from './sample_data/ecommerce_saved_objects';
+import { getFlightsSavedObjects } from './sample_data/flights_saved_objects.js';
+import { getWebLogsSavedObjects } from './sample_data/web_logs_saved_objects.js';
+import { checkLicense } from '../check_license';
+
+export class MapPlugin {
+  setup(core, plugins, __LEGACY) {
+    const { xpackMainPlugin } = __LEGACY.plugins;
+    let routesInitialized = false;
+
+    xpackMainPlugin.registerFeature({
+      id: APP_ID,
+      name: i18n.translate('xpack.maps.featureRegistry.mapsFeatureName', {
+        defaultMessage: 'Maps',
+      }),
+      icon: APP_ICON,
+      navLinkId: APP_ID,
+      app: [APP_ID, 'kibana'],
+      catalogue: [APP_ID],
+      privileges: {
+        all: {
+          savedObject: {
+            all: [MAP_SAVED_OBJECT_TYPE, 'query'],
+            read: ['index-pattern']
+          },
+          ui: ['save', 'show', 'saveQuery'],
+        },
+        read: {
+          savedObject: {
+            all: [],
+            read: [MAP_SAVED_OBJECT_TYPE, 'index-pattern', 'query']
+          },
+          ui: ['show'],
+        },
+      }
+    });
+
+
+    watchStatusAndLicenseToInitialize(xpackMainPlugin, __LEGACY.pluginRef,
+      async license => {
+        if (license && license.maps && !routesInitialized) {
+          routesInitialized = true;
+          initRoutes(__LEGACY, license.uid);
+        }
+      });
+
+    xpackMainPlugin.info
+      .feature(this.id)
+      .registerLicenseCheckResultsGenerator(checkLicense);
```

Comment:
This can all be accessed via `server.newPlatform.setup.licensing`

Reply:
I was able to swap out the top function for:

```
licensing.license$.subscribe(license => {
      const mapsFeature = license.getFeature('spatial');
      if (!routesInitialized && mapsFeature.isEnabled) {
        routesInitialized = true;
        initRoutes(__LEGACY, license.uid);
      }
    });
```
Would love a second set of eyes on it though. I'm _guessing_ spatial refers to maps here since it was the only feature name resembling maps functionality. Need to confirm though.

Wasn't sure what exactly replaces `xpackMainPlugin.info...`, any thoughts?

Reply:
As I can see you didn't use any feature before. Probably you should, but it wasn't the case.
How your plugin works with the LP licensing:
- it requests a feature with `maps` id. https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/maps/index.js#L149-L151
Note, that your plugin uses only `registerLicenseCheckResultsGenerator` and **doesn't** read any feature field.
-  `registerLicenseCheckResultsGenerator` is called with your `checkLicense`.https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/maps/check_license.js
here `checkLisence` receive **xPackInfo** as an argument, not feature!

It means that an equivalent of this logic in the NP will be ~:
```ts
import { LICENSE_CHECK_STATE } from 'license plugin';

licensing.license$.subscribe(license => {
  const { state, message } = license.check('maps', 'basic');
  if (state === LICENSE_CHECK_STATE.Valid && !routesInitialized) {
    routesInitialized = true;
    initRoutes(__LEGACY, license.uid);
  }
});
```

@joshdover I'm thinking should provide state enum as a part of the license to avoid additional import. the check from above becomes:
```js
if (state === license.states.Valid && ...) {
```